### PR TITLE
Implement TCP netconsole.

### PR DESCRIPTION
**Pull request by PeterJohnson**
For 2018, FRC_NetComm provides a TCP netconsole on port 1741.
This is a tagged data stream with both error/warning and console
information.

The console is now colorized similar to the DS console.
The console font and colors are customizable.

TODO:

- ~Add configuration option to turn off color completely~

- [x] Add configuration option (or toolbar button) to show timestamps

- ~Add configuration option (or toolbar button) to hide error message stack traces~

- ~Add UDP listener for backwards compatibility? (TBR)~

### CODE REVIEW

File: edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConnector.java [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+      logger.log("Connected to robot at " + socket.getInetAddress().getHostAddress());
+      return socket;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private void startDsConnect() {
+    startAttempt("DS", new Thread(() -> {
```
Comment (JLLeitschuh) [inline edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConnector.java]: You need to make this a daemon thread or the process won't die when the main thread exits.

File: edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConnector.java [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+        finishAttempt("DS");
+      }
+    }));
+  }
+
+  // Address resolution can take a while, so we do this in a separate
+  // thread, and then kick off a connect attempt on every IPv4 that
+  // resolves.
+  private void startConnect(String host) {
+    startAttempt(host, new Thread(() -> {
```
Comment (JLLeitschuh) [inline edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConnector.java]: Needs to be a daemon thread for same reason as above.

File: edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConnector.java [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+          //logger.log("could not resolve " + host);
+        }
+      } finally {
+        finishAttempt(host);
+      }
+    }));
+  }
+
+  private void startConnect(InetAddress address) {
+    startAttempt(address.getHostAddress(), new Thread(() -> {
```
Comment (JLLeitschuh) [inline edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConnector.java]: Daemon thread

File: edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConsole.java [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+  }
+
+  private static byte[] emptyFrame = new byte[] {0, 0};
+
+  private DataInputStream connect(int team) {
+    logger.log("connecting to team " + team);
+    Socket mySocket;
+    try {
+      mySocket = rioConnector.connect(team);
+    } catch (InterruptedException e) {
```
Comment (JLLeitschuh) [inline edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConsole.java]: Reset the thread's interrupted flag.
`Thread.currentThread().interrupt()`

Explanation:
http://www.yegor256.com/2015/10/20/interrupted-exception.html

File: edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConsole.java [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+    //logger.log("socket connected");
+
+    // kick off keep alive thread
+    sender = new Thread(() -> {
+      while (!Thread.interrupted() && !cleanup.get()) {
+        try {
+          Thread.currentThread().sleep(2);
+          out.write(emptyFrame);
+          out.flush();
+        } catch (InterruptedException e) {
```
Comment (JLLeitschuh) [inline edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConsole.java]: Reset the thread's interrupted flag, see above.

File: edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConsole.java [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+            lock.unlock();
+          }
+          //logger.log("starting reconnect");
+          in = null;
+          Integer teamNumber = teamNumberSupplier.get();
+          if (teamNumber == null) {
+            try {
+              // wait a bit so we don't hammer the CPU
+              Thread.currentThread().sleep(5);
+            } catch (InterruptedException e) {
```
Comment (JLLeitschuh) [inline edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConsole.java]: Thread interrupt flag

**Review by JLLeitschuh — DISMISSED** [review]
Fixup the threading stuff please.

File: edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConsole.java (hunk 5 above)
Comment (PeterJohnson) [reply edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConsole.java]: I'm not seeing a reason to do that here.  We're in the main thread function, there is no caller.

File: edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConsole.java (hunk 6 above)
Comment (PeterJohnson) [reply edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConsole.java]: Ditto

File: edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConnector.java (hunk 3 above)
Comment (PeterJohnson) [reply edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConnector.java]: It is.  See startAttempt().

File: edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConnector.java (hunk 2 above)
Comment (PeterJohnson) [reply edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConnector.java]: It is. See startAttempt().

File: edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConnector.java (hunk 1 above)
Comment (PeterJohnson) [reply edu.wpi.first.wpilib.plugins.riolog/src/netconsole3/RioConnector.java]: It is. See startAttempt().

**Pull request merged**